container/lxd/network.go

```diff
@@ -329,34 +356,94 @@ func ipv6BridgeConfigError(fileName string) error {
+func (s interfaceInfoSlice) Less(i, j int) bool {
+	return s[i].InterfaceName < s[j].InterfaceName
+}
+
+// SortInterfaceInfo sorts a slice of InterfaceInfo on DeviceIndex in ascending
+// order.
+func sortInterfacesByName(interfaces []network.InterfaceInfo) {
+	sort.Sort(interfaceInfoSlice(interfaces))
+}
+
+// errIPV6NotSupported is the error returned by glibc for attempts at
+// unsupported protocols.
 const errIPV6NotSupported = `socket: address family not supported by protocol`
 
-// EnableHTTPSListener configures LXD to listen for HTTPS requests, rather than
-// only via a Unix socket. Attempts to connect over IPv6, but falls back to
-// IPv4 if that has been disabled with in the kernel.
-// Returns an error if updating the server configuration fails.
-func (s *Server) EnableHTTPSListener() error {
-	// Make sure the LXD service is configured to listen to local https
-	// requests, rather than only via the Unix socket.
-	// TODO: jam 2016-02-25 This tells LXD to listen on all addresses,
-	//      which does expose the LXD to outside requests. It would
-	//      probably be better to only tell LXD to listen for requests on
-	//      the loopback and LXC bridges that we are using.
-	if err := s.UpdateServerConfig(map[string]string{
-		"core.https_address": "[::]",
-	}); err != nil {
-		// If the error hints that the problem might be a unsupported protocol,
-		// such as what happens when IPv6 is disabled with in the kernel, we
-		// try IPv4 as a fallback.
-		cause := errors.Cause(err)
-		if strings.HasSuffix(cause.Error(), errIPV6NotSupported) {
-			return errors.Trace(s.UpdateServerConfig(map[string]string{
-				"core.https_address": "0.0.0.0",
-			}))
+// DevicesFromInterfaceInfo uses the input interface info collection to create a
+// map of network device configuration in the LXD format.
+// Names for any networks without a known CIDR are returned in a slice.
+func DevicesFromInterfaceInfo(interfaces []network.InterfaceInfo) (map[string]device, []string, error) {
+	nics := make(map[string]device, len(interfaces))
+	var unknown []string
+
+	for _, v := range interfaces {
+		if v.InterfaceType == network.LoopbackInterface {
+			continue
 		}
-		return errors.Trace(err)
+		if v.InterfaceType != network.EthernetInterface {
+			return nil, nil, errors.Errorf("interface type %q not supported", v.InterfaceType)
+		}
+		if v.ParentInterfaceName == "" {
+			return nil, nil, errors.Errorf("parent interface name is empty")
+		}
+		if v.CIDR == "" {
+			unknown = append(unknown, v.ParentInterfaceName)
+		}
+		nics[v.InterfaceName] = newNICDevice(v.InterfaceName, v.ParentInterfaceName, v.MACAddress, v.MTU)
 	}
-	return nil
+
+	return nics, unknown, nil
+}
+
+// newNICDevice creates and returns a LXD-compatible config for a network
+// device from the input arguments.
+func newNICDevice(deviceName, parentDevice, hwAddr string, mtu int) device {
+	device := map[string]string{
+		"type":    "nic",
+		"nictype": "bridged",
```

Comment:
I thought we were looking to support "bridged" or "macvlan" (at least as allowed in the default config).
I imagine you have to match the nictype to the parent object?